garage/misc/git_handler.py

```diff
@@ -0,0 +1,382 @@
+        self.target_branch_name = ("run_exp_on_" + str(int(time.time())))
+        try:
+            self._switch_to_target(self.target_branch_name, self.target_sha)
+            current_sha = self.get_sha()
+            self._run_experiment(exp_args)
+        except BaseException:
+            print(colored(
+                "If the working area is damaged after " \
+                "an unexpected error while trying to\nswitch to the " \
+                "target in the repository, run the following command(s) " \
+                "to\nrestore it:", "yellow"))
+            if self.branch_name:
+                print(
+                    colored("$ git checkout %s" % (self.branch_name),
+                            "yellow"))
+            else:
+                print(
+                    colored("$ git reset --hard %s" % (self.restore_sha),
+                            "yellow"))
+
+            if self.stash_sha:
+                print(
+                    colored("$ git stash apply %s" % (self.stash_sha),
+                            "yellow"))
+            raise
+        finally:
+            self._restore_working_dir(self.target_branch_name, self.stash_sha,
+                                      self.branch_name, self.restore_sha)
+
+    def _verify_target(self, target):
+        """Verify the key-value target is valid.
+
+        Parameters
+        ----------
+        - target: same description as in method run_experiment_on_target.
+
+        Returns
+        -------
+        If the target is valid, the corresponding SHA.
+
+        """
+        target = [target.strip() for target in target.split(":")]
+        sha = ""
+        assert (len(target) == 2 and target[0]
+                and target[1]), self._get_target_use_message()
+        try:
+            verify = self.valid_targets[target[0]]
+            sha = verify(target[1])
+        except KeyError:
+            print(self._get_target_use_message())
+            raise
+        return sha
+
+    def _get_target_use_message(self):
+        """Print a use message if the user passed an invalid target."""
+        return colored(
+            "\nInvalid target: targets can be defined with a SHA, branch " \
+            "name (local or remote) or tag\nas shown " \
+            "in the following examples:\nsha: 8d54s23a" \
+            "\nbranch: my_branch\nbranch: remote/master\ntag: my_tag\n",
+            "yellow")
+
+    def _switch_to_target(self, target_branch_name, target_sha):
+        """Clean working directory and check out the target branch.
+
+        The target branch is created during the checkout operation.
+        SIGINT is temporarily blocked while performing the git operations to
+        avoid corrupting the working directory.
+        """
+        # Block SIGINT to avoid ending in an unknown state
+        signal.pthread_sigmask(signal.SIG_BLOCK, {signal.SIGINT})
+        # Clean working directory
+        self.reset()
```

Comment:
Should we hard reset here before switching to target branch? If a launcher file is written and not commit, we will lose it with this reset right? I think a more proper way is throwing an error to guide the user to either stash the changes or commit it.

Reply:
Sure Zhanpeng, I create a stash before reseting the working directory [here](https://github.com/rlworkgroup/garage/blob/a384e0a469d953353fac4cb210e9db30d6d0849a/garage/misc/git_handler.py#L205), and once the environment is restored, the stash is applied [here](https://github.com/rlworkgroup/garage/blob/a384e0a469d953353fac4cb210e9db30d6d0849a/garage/misc/git_handler.py#L327-L328).
I also print the sha of the stash that was created (if any unstaged changes were found) [here](https://github.com/rlworkgroup/garage/blob/a384e0a469d953353fac4cb210e9db30d6d0849a/garage/misc/git_handler.py#L227-L230), so the users are able to apply the stash if anything went wrong.